src/Hamiltonians/HOCartesian2BosonRelative.jl

```diff
@@ -0,0 +1,230 @@
+```math
+    V_{\\mathbf{ij}} = \\prod_{d \\in x, y,\\ldots} \\psi_{i_d}(0) \\psi_{j_d}(0).
+```
+Only even parity states feel this interaction, so all ``i_d`` are even. Note that the matrix 
+representation of this Hamiltonian is completely dense in the even-parity subspace.
+
+# Arguments
+
+* `addr`: The starting address is required for the [`Hamiltonians`](@ref) interface,
+    but is overwritten for this Hamiltonians. A warning will be issued if the input address 
+    has the wrong number of particles (1) or modes (`prod(S)`). This warning can be disabled 
+    with keyword argument `warn_address = false`.
+* `S`: Tuple of the number of levels in each dimension, including the groundstate. The first 
+    dimension should be the largest and sets the energy scale of the system, ``\\hbar\\omega_x``.
+    The aspect ratios are determined from the other elements of `S`.
+    Defaults to a 1D spectrum with number of levels matching modes of `addr`.
+* `Sx`, `ηs`: Alternatively, provide the number of modes (including the groundstate) in the 
+    first dimension `Sx` and `D-1` aspect ratios for the other dimensions `ηs = (η_y, ...)`. 
+    The elements of `ηs` should be at least `1.0`. Providing `Sx` and `ηs` will redefine 
+    `S = (Sx, Sy, ...)`.
+* `g`: the (bare) isotropic interparticle interaction parameter. The value of `g` is assumed 
+    to be in trap units.
+* `interaction_only`: if set to `true` then the noninteracting single-particle terms are 
+    ignored. Useful if only energy shifts due to interactions are required.
+
+See also [`HOCartesianEnergyConserved`](@ref) and[`HOCartesianEnergyConservedPerDim`](@ref).
+"""
+struct HOCartesian2BosonRelative{
+    D,  # number of dimensions
+    A<:BoseFS
+} <: AbstractHamiltonian{Float64}
+    addr::A
+    S::NTuple{D,Int64}
+    aspect::NTuple{D,Float64}
+    vtable::Vector{Float64}  # interaction coefficients
+    g::Float64
+    interaction_only::Bool
+end
+
+function HOCartesian2BosonRelative(
+        addr; # ignored for this Hamiltonian if no other arguments are provided
+        S = (num_modes(addr),),
+        Sx = nothing,
+        ηs = nothing, 
+        g = 1.0,
+        interaction_only = false,
+        warn_address = true
+    )
+    if isnothing(ηs) && isnothing(Sx)
+        S[1] ≠ maximum(S) && throw(ArgumentError("First dimension must have the most states"))
+        aspect = float.(box_to_aspect(S))
+        D = length(S)
+    else
+        any(ηs .< 1.0) && throw(ArgumentError("Aspect ratios must be greater than 1.0"))
+        D = length(ηs) + 1
+        Srest = map(x -> floor(Int, (Sx - 1) / x + 1), ηs)
+        S = (Sx, Srest...)
+        aspect = (1.0, float.(ηs)...)
+    end
+    P = prod(S)
+
+    # address type is fixed so it is overwritten here to ensure it works
+    warn_address && !(addr isa BoseFS{1,P}) && @warn "Input address is being overwritten"
+    addr = BoseFS(P, 1 => 1)
```

Comment:
I suggest replacing this by code that checks whether `addr` and `S` are compatible, like you did for the other HO Hamiltonians.